OSSMDOC-395: Configure Federation for failover. (#41725)
openshift/openshift-docs: modules/ossm-federation-config-failover.adoc

////
This module included in the following assemblies:
* service_mesh/v2x/ossm-federation.adoc
////

> :_content-type: PROCEDURE ????????????????

Author: Done

[id="ossm-federation-config-failover_{context}"]

> = Configuring a federated mesh for failover

Reviewer: ```suggestion
= Configure a federated mesh for failover
```
To use in a concept file I would drop the "ing".

Failover is the ability to switch automatically and seamlessly to a reliable backup system, for example another server. In the case of a federated mesh, you can configure a service in one mesh to failover to a service in another mesh.

You configure Federation for failover by setting the `importAsLocal` and `locality` settings in a `ServiceImportSet` resource and then configuring a `DestinationRule` that configures failover for the service to the locality specified in the `ServiceImportSet`.

.Prerequisites

> * Two or more OpenShift Container Platform 4.6 or above clusters already networked and federated.

Reviewer: ```suggestion
* Two or more {product-title} 4.6 or above clusters already networked and federated.
```

* `ExportServiceSet` resources already created for each mesh peer in the federated mesh.
* `ImportServiceSet` resources already created for each mesh peer in the federated mesh.
* An account with the cluster-admin role.

> == Configuring an ImportServiceSet for failover

Reviewer: I would put `==Configuring an ImportServiceSet for failover `into a new module and use that module as a procedure module. You cannot have multiple headings in a procedure file.

In the examples in this section, the `green-mesh` is located in the `us-east` region, and the `red-mesh` is located in the `us-west` region.

.Example ImportServiceSet resource from red-mesh to green-mesh
[source,yaml]
----
kind: ImportedServiceSet
apiVersion: federation.maistra.io/v1
metadata:
  name: red-mesh #name of mesh that exported the service
  namespace: green-mesh-system #mesh namespace that service is being imported into
spec:
  importRules: # first matching rule is used
  # import ratings.bookinfo as ratings.bookinfo
  - type: NameSelector
    importAsLocal: true
    nameSelector:
      namespace: bookinfo
      name: ratings
      alias:
        # service will be imported as ratings.bookinfo.svc.red-mesh-imports.local
        namespace: bookinfo
        name: ratings
  #Locality within which imported services should be associated.
  locality:
    region: us-west
----

> .Table ImportedServiceLocality Fields

Reviewer: ```suggestion
.`ImportedServiceLocality` fields table
```
Add an introduction sentence here to introduce this table.

|===
| Name | Description | Type

|region:
|Region within which imported services are located.
|string

|subzone:
|Subzone within which imported services are located.  I Subzone is specified, Zone must also be specified.
|string

|zone:
|Zone within which imported services are located.  If Zone is specified, Region must also be specified.
|string
|===

.Procedure

> . Log in to the {product-title} CLI as a user with the `cluster-admin` role. Enter the following command.

Reviewer: ```suggestion
. Log in to the {product-title} CLI as a user with the `cluster-admin` role, enter the following command:
```

> https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

+
[source,terminal]
----
$ oc login --username=<NAMEOFUSER> <API token> https://<HOSTNAME>:6443
----
+

> . Change to the project where you installed the control plane.

Reviewer: ```suggestion
. Change to the project where you installed the control plane, enter the following command:
```

+
[source,terminal]
----
$ oc project <smcp-system>
----
+
For example, `green-mesh-system`.
+
[source,terminal]
----
$ oc project green-mesh-system
----
+

> . Run the following command to edit the `ImportServiceSet` file, where `<ImportServiceSet.yaml>` includes a full path to the file you want to edit:

Reviewer: To maintain some consistency with the above procedure maybe rework this to include ",enter the following command"? Something like this?

> . Run the following command to edit the `ImportServiceSet` file, where `<ImportServiceSet.yaml>` includes a full path to the file you want to edit:

Reviewer: ```suggestion
. Edit the `ImportServiceSet` file, where `<ImportServiceSet.yaml>` includes a full path to the file you want to edit, enter the following command:
```

+
[source,terminal]
----
$ oc edit -n <smcp-system> -f <ImportServiceSet.yaml>
----
+
For example, if you want to modify the file that imports from the red-mesh-system to the green-mesh-system as shown in the previous `ImportServiceSet` example.
+
[source,terminal]
----
$ oc edit -n green-mesh-system -f import-from-red-mesh.yaml
----
. Modify the file:
.. Set `spec.importRules.importAsLocal` to `true`.
.. Set `spec.locality` to a `region`, `zone`, or `subzone`.
.. Save your changes.

//Anything else we need to do to deploy/update the changed file?

> == Configuring a DestinationRule for failover

Reviewer: I would put `==Configuring a DestinationRule for failover` into a new module and use that module as a procedure module.  You cannot have multiple headings in a procedure file.

* Failover policy between regions. This ensures that failover beyond a region boundary will behave predictably.

.Procedure

. Log in to the {product-title} CLI as a user with the `cluster-admin` role. Enter the following command. Then, enter your username and password when prompted.
+
[source,terminal]
----
$ oc login --username=<NAMEOFUSER> <API token> https://<HOSTNAME>:6443
----
+
. Change to the project where you installed the control plane.
+
[source,terminal]
----
$ oc project <smcp-system>
----
+
For example, `green-mesh-system`.
+
[source,terminal]
----
$ oc project green-mesh-system
----
+
. Create a `DestinationRule` file based on the following example where if green-mesh is unavailable, the traffic should be routed from the green-mesh in the `us-east` region to the red-mesh in `us-west`.

.Example DestinationRule
[source,yaml]
----
apiVersion: networking.istio.io/v1beta
kind: DestinationRule
metadata:
  name: default
  namespace: green-mesh-system
spec:
  host: "*.green-mesh-system.svc.cluster.local"
  trafficPolicy:
    loadBalancer:
      localityLbSetting:
        enabled: true
        failover:
          - from: us-east

>           to: us-west

Reviewer 2: @JStickler There is a syntax indentation error in line 161.
The `to` should align with `from` . Need to add two more spaces.
Reference:
https://istio.io/latest/docs/tasks/traffic-management/locality-load-balancing/failover/#configure-locality-failover

    outlierDetection:
      consecutive5xxErrors: 1

>       interval: 1s

Author: Updated the value.

      baseEjectionTime: 1m
----

> . Run the following command to deploy the `DestinationRule`, where `<DestinationRule>` includes the full path to your file.

Reviewer: ```suggestion
. Deploy the `DestinationRule`, where `<DestinationRule>` includes the full path to your file, enter the following command:
```
